### ISSUE TITLE
Feature request: Ability to delegate crypto operations for stub keys

### ISSUE THREAD

**Issue by tadfisher** [vanitasvitae/pgpainless#318]
I'm working on [adding hardware security key support to Android Password Store](https://github.com/android-password-store/Android-Password-Store/discussions/2129), and a particular pain point with the integration is dealing with stub encryption keys that are offloaded to hardware.

PGPainless does expose the presence of the S2K dummy value, which can be queried for individual keys. From a cursory perusal, it doesn't offer an easy way to delegate to client code when encountering such a key during encrypt/decrypt operations.

I was thinking of proposing callbacks such as:

```java
public interface StubKeyDecryptionCallback {
    @Nullable InputStream decrypt(KeyInfo keyInfo, ConsumerOptions options, InputStream inputStream);
}

public interface StubKeyEncryptionCallback {
    @Nullable OutputStream encrypt(KeyInfo keyInfo, ProducerOptions options, OutputStream outputStream);
}
```

But I don't know the ramifications of trying to wrap an externally-generated stream in the implementation.

In addition, while the presence of GNU_DUMMY_S2K is a good indicator of a hardware-offloaded key, there's an [additional S2K extension](https://github.com/gpg/gnupg/blob/master/doc/DETAILS#gnu-extensions-to-the-s2k-algorithm) (colloquially `gnu-divert-to-card`) that includes the card serial number, which is would also be useful but not strictly necessary.

If there's a mechanism that allows for this without modifying PGPainless, any advice would be appreciated!

**Comment by tadfisher** [comment]
Forgive me, I didn't have enough coffee this morning. I don't believe delegating **encryption** would be necessary as that doesn't require the secret key.

**Comment by vanitasvitae** [comment]
Hey!
Having support for keys on hardware would be an awesome feature indeed. I have no idea how to implement that using Bouncycastle though, so some research needs to be done :D

This is not on my immediate roadmap though, but I do of course welcome any pointers, resources, pull requests etc. :)

**Comment by tadfisher** [comment]
Okay, now that I understand a little more about how PGP works, it looks like all the card protects is the session key. So a callback that delegates `decryptSessionKey` when encountering that GNU_DUMMY_S2k value should be sufficient. The library I'm using ([hwsecurity](https://github.com/android-password-store/hwsecurity)) only expects the raw encrypted key as a byte array.

**Comment by vanitasvitae** [comment]
Interesting, I haven't yet looked into hwsecurity, but it appears to be worth doing so! :D

One issue I see is that hwsecurity seems to be bound to Android, which PGPainless is not. Therefore we cannot depend on it.
That means that the best way to go about is probably to create some kind of callback the user can register to handle dealing with hwsecurity themselves.
Such a callback could receive the Public Key Encrypted Session Key Packet and return the decrypted session key for PGPainless to try decryption with.
This sounds doable :)

**Comment by tadfisher** [comment]
> Such a callback could receive the Public Key Encrypted Session Key Packet and return the decrypted session key for PGPainless to try decryption with.

That would be perfect, thank you!

**Comment by vanitasvitae** [comment]
Do you happen to know, what information a downstream application (e.g. Android Password Store) has when setting up decryption? I.e. does the application know beforehand, which keys are available on smartcards/tokens?

I took a look at how OpenKeychain handle this and they appear to also maintain a caching layer for decrypted session keys. Implementing a similar caching layer may make sense eventually.

**Comment by vanitasvitae** [comment]
@tadfisher could you do me a favor and try out PGPainless built from the [hardwareDecryption](https://github.com/pgpainless/pgpainless/tree/hardwareDecryption) branch?

There is a [test case](https://github.com/pgpainless/pgpainless/blob/hardwareDecryption/pgpainless-core/src/test/java/org/pgpainless/decryption_verification/CustomPublicKeyDataDecryptorFactoryTest.java#L38-L85) which demonstrates how to use the new API.

You would need to change [these lines](https://github.com/pgpainless/pgpainless/blob/hardwareDecryption/pgpainless-core/src/test/java/org/pgpainless/decryption_verification/CustomPublicKeyDataDecryptorFactoryTest.java#L60-L68) to integrate the hardware SDK for session key decryption.

Let me know if the new API suites your needs, and what error cases are there to accommodate with subtypes of the HardwareSecurityException.

**Comment by vanitasvitae** [comment]
I further created #321 which might be useful here.

**Comment by msfjarvis** [comment]
Unfortunately I'm quite short on time for the next few days so I can't test this myself, but the proposed API looks great to me. Thanks for the extremely quick turnaround on this issue!

**Comment by tadfisher** [comment]
Thank you, I'll take a look now.

**Comment by tadfisher** [comment]
> Do you happen to know, what information a downstream application (e.g. Android Password Store) has when setting up decryption? I.e. does the application know beforehand, which keys are available on smartcards/tokens?

There are a few constraints we need to deal with:

1. Smartcards have slots for signing, authentication, and encryption subkeys. While it is possible to store a primary key in all three slots, it's not easy to do so with existing tools; usually the user generates them and "moves" them to the card (creating stub keys in their keychain), or generates them on the card itself.
2. Smartcards have an optional field for the preferred URL of the primary public key, but there is no canonical place to store even a key ID. So the general pairing process is:
  - Import public key+subkeys into keychain
  - Open the smartcard via USB or NFC to get the card ID and the IDs of its subkeys
  - Replace the keychain's public keys with "stub" keys; the primary key gets an S2K value that says, "there is no secret key data here", and the subkeys get S2K values that say, "the secret key is stored on a card with this ID". These are the "GNU S2K extensions" as described [here](https://github.com/gpg/gnupg/blob/master/doc/DETAILS#gnu-extensions-to-the-s2k-algorithm).
    - We could also choose to forgo the GnuPG "standard" and keep track of these keys separately, if we don't have to be compatible with exported GnuPG keychains.

Once we have that info, then we can perform decryption by:

1. Recognizing that the encryption subkey for a message's recipient is diverted to a smartcard
2. Prompting the user to enter a PIN and insert/tap the corresponding card
3. Immediately sending the encrypted session key through the transport layer to the card, which spits out the decrypted session key
4. Using that session key to decrypt the message

Another option is to have a setting that says, "always try decryption with a smartcard", and simply attempt steps 2-4 above for each decryption.

So to answer your question, we _can_ know which keys are available beforehand, or we could choose to try decryption with any card the user connects to the host.

> I took a look at how OpenKeychain handle this and they appear to also maintain a caching layer for decrypted session keys. Implementing a similar caching layer may make sense eventually.

Agreed, though I don't know that this works for smartcards. At least with GnuPG and OpenKeychain, one must have the card available for each decryption. They both cache the card's PIN though.

**Comment by vanitasvitae** [comment]
> At least with GnuPG and OpenKeychain, one must have the card available for each decryption.

I think this is about very-short-term caching only (recreation of Activities for example).

**Comment by tadfisher** [comment]
I at least have the callback successfully invoked and the hwsecurity UI appearing. Unfortunately, something is wrong with the decryption command; most likely the key data needs to be formatted differently. I'll investigate more, but this looks promising so far.

**Comment by vanitasvitae** [comment]
Let me know if I can assist you in any way.
Feel free to provide feedback for the API, so that we can merge a suitable interface into the main branch eventually :)

**Comment by tadfisher** [comment]
Success! PGPainless has decrypted payloads with an RSA secret key residing on a Yubikey, over both NFC and USB.

ECDH keys will take more work, because we require more information (the symmetric key size and some sort of "user keying data", which is unknown to me).

Right now the DecryptionCallback runs twice, because it is called for the main key as well as the encryption subkey; I believe this line in `getIdsOfHardwareBackedKeys` is causing the problem.

```java
// TODO: Is GNU_DUMMY_S2K appropriate?
if (type == S2K.GNU_PROTECTION_MODE_DIVERT_TO_CARD || type == S2K.GNU_DUMMY_S2K) {
    hardwareBackedKeys.add(secretKey.getKeyID());
}
```

There are two meanings to the `GNU_DUMMY_S2K` type, which are distinguished by the protection mode:

- 1: No secret key exists (no flag for this in BouncyCastle)
- 2: The secret key exists on a smartcard (`S2K.GNU_PROTECTION_MODE_DIVERT_TO_CARD`)

So you must check *both* of these conditions:

```java
int type = s2K.getType();
int mode = s2K.getProtectionMode();
if (type == S2K.GNU_DUMMY_S2K && mode == S2K.GNU_PROTECTION_MODE_DIVERT_TO_CARD) {
    ...
}
```

**Comment by tadfisher** [comment]
Ah, that's not the reason the callback is run twice. In `DecryptionStreamFactory#decryptSessionKey`:

```java
InputStream decryptedDataStream = publicKeyEncryptedData.getDataStream(decryptorFactory);
PGPSessionKey pgpSessionKey = publicKeyEncryptedData.getSessionKey(decryptorFactory);
```

`getDataStream` internally calls `getSessionKey`, so I think BouncyCastle wants you to construct a `SessionKeyDataDecryptorFactory` with the results of `getSessionKey` to avoid calling `recoverSessionData()` twice.

**Comment by vanitasvitae** [comment]
Woah thats so awesome!

Yeah, the `SessionKeyDataDecryptorFactory` did not exist until recently when I proposed it as a pull request to the BC dev team ;)
I haven't yet gotten around to update my code to actually make use of it in PGPainless code.

Alterantively, as a workaround you could wrap your `DecryptorFactory` in a [CachingPublicKeyDataDecryptorFactory](https://github.com/pgpainless/pgpainless/pull/321), so that the double-call only is forwarded to the card once.

> ECDH keys will take more work, because we require more information (the symmetric key size and some sort of "user keying data", which is unknown to me).

Is it possible, that this data is at array index 1 [here](https://github.com/pgpainless/pgpainless/pull/322/files#diff-00a6fb91b550c48c7b020b27d2e2616970df629a035ce2b859ee4b3a28d7bb2fR91)? I think at least for ElGamal keys, the key is split into two parts, but peaking into the OpenKeychain code, they appeared to always only pass the index-0 data to the card.

**Comment by wiktor-k** [comment]
>  ECDH keys will take more work, because we require more information (the symmetric key size and some sort of "user keying data", which is unknown to me).

AFAIK ECDH decryption requires data that's in user certificate, like KDF (so it cannot operate in principle without user's certificate/public key). I believe @hko-s had some heuristics that try to avoid this issue but I'm not sure about details...

**Comment by vanitasvitae** [comment]
> I believe @hko-s had some heuristics that try to avoid this issue but I'm not sure about details...

This work can probably be found somewhere in here: https://gitlab.com/openpgp-card

**Comment by wiktor-k** [comment]
Btw since Java has somewhat standardized crypto API maybe it would be possible to utilize that? This would allow signing and decryption using PKCS#11 cards out of the box... Maybe that's not in scope for now but worth keeping in mind.

For the record the following conversations on `#sequoia` (OFTC) may also be relevant:

```
12:25:58 - Valodim: ah, that makes sense. well, ripping that stuff out from hwsecurity should be fairly doable, it's modularized so just gotta cut away the unnecessary parts 😁
12:47:12 - muppeth has left the room (Quit: Remote host closed the connection).
12:52:43 - vanitasvitae: Valodim: isn't hwsecurity published under a commercial license?
12:53:01 - Valodim: also under GPL
12:53:13 - vanitasvitae: I see.
12:54:53 - vanitasvitae: PGPainless is licensed as Apache 2.0, so the code-ripping would need to be done elsewhere 😂
12:55:19 - vanitasvitae: But apparently glueing on hwsecurity onto PGPainless is doable, as demonstrated 😂
13:34:45 - wiktor: alternatively have interfaces in pgpainless and a gpl-licensed project that implements them using hwsecurity sdk
13:36:59 - vanitasvitae: Hehe, sounds like a potential for cooperation PGPainless x Cotech 😉
13:38:16 - vanitasvitae: wiktor, thanks for hinting to the work by heiko[m] in the thread!
```

**Comment by hko-s** [comment]
> AFAIK ECDH decryption requires data that's in user certificate, like KDF (so it cannot operate in principle without user's certificate/public key). I believe @hko-s had some heuristics that try to avoid this issue but I'm not sure about details...

Right, using ECC keys from an OpenPGP card requires information that is not available on the OpenPGP card itself. Historically, software has required users to provide the OpenPGP public key to obtain that data.

There is a lengthy discussion of the topic, here: https://gitlab.com/openpgp-card/openpgp-card/-/issues/2

For experimentation, I implemented a hack that brute-forces the missing bit of information: [openpgp-card-sequoia/src/util.rs#L166-226](https://gitlab.com/openpgp-card/openpgp-card/-/blob/4dfaadc5adb02f545a25974015ed436b11fae497/openpgp-card-sequoia/src/util.rs#L166-226)
The approach relies on the fact that the "unknown" (to the card) parameters are hashed into the PGP fingerprint, which *is* available on an OpenPGP card. So one can guess the correct parameters and then verify the guess against the fingerprint.

(The code above uses a list of the most common parameters, but the space of possible parameters is small, and could just as well be fully searched)

**Comment by tadfisher** [comment]
@wiktor-k The actual `javax.smartcardio` API isn't available on Android, but there are a couple of backport projects out there on GitHub. The implementation would have to be completely different on Android because the OpenJDK version uses pcsclite on Linux, which definitely won't work out-of-the-box on Android.

(Referring to Android here because I'm working on smartcard support for Android-Password-Store)

**Comment by vanitasvitae** [comment]
I guess there won't be much shared code between a solution based on hwsecurity and one based on javax.smartcardio.
Its probably a sensible idea to go with hwsecurity for now and create an implementation based on that.

I believe it should be possible to create some implementations of BC classes such as `PublicKeyDataDecryptorFactory` which delegate to the smartcard. We should also look into signature verification and possibly authentication (never looked into that so far).

That way we get a solution which is compatible to PGPainless, but which can also used with plain BC.
PGPainless can do stuff like certificate evaluation on top of that, just like it already does.

At a later point we can (if there is demand) create a similar implementation using javax.smartcardio.

**Comment by wiktor-k** [comment]
Thanks for the background info @tadfisher really appreciate that :bow: (and also the great work you're putting to get first-class smartcards support).

I didn't mention the fact that strictly speaking the support for parsing GnuPG card stubs is neither necessary nor sufficient for decryption with smartcards. The public bits in the certificate (public keys) are enough (and only used for ECDH). I believe this is how OpenKeychain works too.

@vanitasvitae Just as a couple of pointers here's the interfaces that Sequoia exposes: [`Signer`](https://docs.sequoia-pgp.org/sequoia_openpgp/crypto/trait.Signer.html) and [`Decryptor`](https://docs.sequoia-pgp.org/sequoia_openpgp/crypto/trait.Decryptor.html) (as for decryptor I'd rather see two methods: one for RSA decryption that's direct and one for EC shared key derivation). They are quite simple and you could expose similar interfaces in PGPainless that they would/could be implemented by hwsecurity or anything else.

Not sure about the async nature since decrypting with smartcards would probably need user's PIN and that would block the decryption... Just some food for thought I hope you don't mind :)

**Comment by tadfisher** [comment]
> AFAIK ECDH decryption requires data that's in user certificate, like KDF (so it cannot operate in principle without user's certificate/public key). I believe @hko-s had some heuristics that try to avoid this issue but I'm not sure about details...

In that case, @vanitasvitae, is it possible to provide at least the associated key ID in `HardwareSecurity.DecryptionCallback`?

**Comment by vanitasvitae** [comment]
> In that case, @vanitasvitae, is it possible to provide at least the associated key ID in HardwareSecurity.DecryptionCallback?

That is possible, yes. I will see what I can do.

**Comment by vanitasvitae** [comment]
I made some changes.
Unfortunately, the actual decryption methods of the PublicKeyDataDecryptorFactory do not pass on the keyId, so I had to create a custom wrapper interface, which makes the keyId accessible.
This sadly also means, that users would need to register one decryptorFactory instance per (sub)key, which might be a minor inconvenience.

**Comment by tadfisher** [comment]
In that case, the custom wrapper isn't necessary, because I can just register a factory per keyId. But this makes the API a bit more straightforward.

**Comment by vanitasvitae** [comment]
As a heads up: I'm planing to do a 1.4.0 release soon, which will include a brand new decryption stream, along with support for hardware tokens for decryption.

Let me know, if you are happy with the current API (https://github.com/pgpainless/pgpainless/blob/pda/pgpainless-core/src/test/java/org/pgpainless/decryption_verification/CustomPublicKeyDataDecryptorFactoryTest.java#L56-L85), or if you'd like to see some changes :)

Among others, `DecryptorFactories` are now identified by `SubkeyIdentifier` (fingerprint + key-id).

**Comment by vanitasvitae** [comment]
I just released PGPainless `1.4.0-rc1` which introduces support for custom decryptor factories :)
